internal/notifier/google_pubsub_topic.go

```diff
@@ -0,0 +1,131 @@
+limitations under the License.
+*/
+
+package notifier
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"cloud.google.com/go/pubsub"
+	"google.golang.org/api/option"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	eventv1 "github.com/fluxcd/pkg/apis/event/v1beta1"
+)
+
+type (
+	// GooglePubSubTopic holds a Google Pub/Sub client and target topic.
+	GooglePubSubTopic struct {
+		topicName string
+
+		client interface {
+			Publish(ctx context.Context, eventPayload []byte) (serverID string, err error)
+		}
+	}
+
+	googlePubSubClient struct {
+		jsonCreds []byte
+		projectID string
+		topicID   string
+		attrs     map[string]string
+	}
+)
+
+// ensure *GooglePubSubTopic implements Interface.
+var _ Interface = &GooglePubSubTopic{}
+
+// NewGooglePubSubTopic creates a Google Pub/Sub client tied to a specific
+// project and topic.
+//
+// The jsonCreds parameter is optional, and if len(jsonCreds) == 0 then the
+// automatic authentication methods of the Google libraries will take place,
+// and therefore methods like Workload Identity will be automatically attempted.
+//
+// The attrs paramter is optional, and if len(attrs) == 0 then no attributes will
+// be added to the Pub/Sub message.
+func NewGooglePubSubTopic(projectID, topicID, jsonCreds string, attrs map[string]string) (*GooglePubSubTopic, error) {
+	if projectID == "" {
+		return nil, errors.New("projectID cannot be empty")
+	}
+	if topicID == "" {
+		return nil, errors.New("topicID cannot be empty")
+	}
+	if len(attrs) == 0 {
+		attrs = nil
+	}
+	return &GooglePubSubTopic{
+		topicName: fmt.Sprintf("projects/%s/topics/%s", projectID, topicID),
+		client: &googlePubSubClient{
+			jsonCreds: []byte(jsonCreds),
+			projectID: projectID,
+			topicID:   topicID,
+			attrs:     attrs,
+		},
+	}, nil
+}
+
+// Post posts Flux events to a Google Pub/Sub Topic.
+func (g *GooglePubSubTopic) Post(ctx context.Context, event eventv1.Event) error {
+	// Skip Git commit status update event.
+	if event.HasMetadata(eventv1.MetaCommitStatusKey, eventv1.MetaCommitStatusUpdateValue) {
+		return nil
+	}
+
+	eventBytes, err := json.Marshal(event)
+	if err != nil {
+		return fmt.Errorf("error json-marshaling event: %w", err)
+	}
+
+	serverID, err := g.client.Publish(ctx, eventBytes)
+	if err != nil {
+		return fmt.Errorf("error publishing event to topic %s: %w", g.topicName, err)
+	}
+
+	// debug log
+	log.FromContext(ctx).V(1).Info("Event published to Google Pub/Sub Topic",
+		"topic", g.topicName,
+		"server message id", serverID)
+
+	return nil
+}
+
+func (g *googlePubSubClient) Publish(ctx context.Context, eventPayload []byte) (serverID string, err error) {
+	var opts []option.ClientOption
+	if len(g.jsonCreds) > 0 {
+		opts = append(opts, option.WithCredentialsJSON(g.jsonCreds))
+	}
+	var client *pubsub.Client
+	client, err = pubsub.NewClient(ctx, g.projectID, opts...)
+	if err != nil {
+		return
+	}
+	defer func() {
+		if closeErr := client.Close(); err == nil {
+			err = closeErr
+		}
```

Comment:
There was an error in the `smoketest` job, in the `Smoke test Fuzzers` step:

https://github.com/fluxcd/notification-controller/actions/runs/5270174993/jobs/9529323364?pr=543
```
Error: internal/notifier/google_pubsub_topic.go:120:17: undefined: errors.Join
```

Which is weird because I can see `go version go1.20.5 linux/amd64` in the output of `Setup Go` :thinking:

Reply:
Ah it's because `make fuzz-smoketest` uses a docker image: `local-fuzzing:latest`

Reply:
But from the `tests/fuzz/Dockerfile.builder` Dockerfile I can't tell if it's downloading a different Go version or if it's supposed to use the host Go installation which was set up by `Setup go` :thinking: